internal/rest/client/client.go

```diff
@@ -312,6 +318,54 @@ func (c *Client) QueryStruct(ctx context.Context, method string, endpointType ty
 	return nil
 }
 
+// Websocket dials the provided endpoint and tries to upgrade the connection.
+//
+// The final URL is that provided as the endpoint combined with the applicable prefix for the endpointType and the scheme and host from the client.
+func (c *Client) Websocket(ctx context.Context, endpointType types.EndpointPrefix, endpoint *api.URL) (*websocket.Conn, error) {
+	// Merge the provided URL with the one we have for the client.
+	localURL := c.mergeURL(endpointType, endpoint)
+
+	// Pick the right scheme based on the client configuration.
+	if c.url.URL.Scheme == "http" {
+		localURL.URL.Scheme = "ws"
+	} else {
+		localURL.URL.Scheme = "wss"
+	}
+
+	// Get the transport configuration from the HTTP client.
+	tr, ok := c.Client.Transport.(*http.Transport)
+	if !ok {
+		return nil, fmt.Errorf("Invalid underlying client transport, expected %T, got %T", &http.Transport{}, c.Client.Transport)
+	}
+
+	// Setup a new websocket dialer using the already existing HTTP client.
+	// As the client might be either local or remote always copy the relevant TLS config too.
+	dialer := websocket.Dialer{
+		NetDialContext:    tr.DialContext,
+		NetDialTLSContext: tr.DialTLSContext,
+		TLSClientConfig:   tr.TLSClientConfig,
+		Proxy:             tr.Proxy,
+		HandshakeTimeout:  time.Second * 5,
```

Comment:
Instead of hard-coding this timeout, I think it might be more configurable to use the timeout from the context, and then fallback to a default if the context has no timeout.

In `rawQuery`, we check that a context has a deadline, and if not we apply a default 30s timeout. I'm not sure if something like that would be useful here as well:
```go
// Assign a context timeout if we don't already have one.
	_, ok := ctx.Deadline()
	if !ok {
		timeoutCtx, cancel := context.WithTimeout(ctx, 30*time.Second)
		ctx = timeoutCtx
		defer cancel()
	}
```

Reply:
I like the idea of adding the timeout to the context if a deadline doesn't yet exist.

To be consistent with the timeout duration I have copied the one from the LXD client https://github.com/canonical/lxd/blob/main/client/lxd.go#L423.
But in LXD we don't use `dialer.DialContext` but just `dialer.Dial`. 
Interestingly in gorilla they apply the [exact same pattern](https://github.com/gorilla/websocket/blob/main/client.go#L241) in case `HandshakeTimeout` is set.

So by setting the deadline ourselves we can omit setting the `HandshakeTimeout` field.

Reply:
Fixed.

Reply:
Oh, one thing that occurred to me is that by using `DialContext` with a 30s default context timeout, that means this websocket will only live for a maximum of 30s, right?

Would it be better in this case to require that the context _must have_ a deadline, rather than setting a default? I'm assuming since this is a generic websocket client helper, there may be use-cases for longer-lived websockets.

Reply:
The context provided for `DialContext` only covers the dialing phase (sending the HTTP upgrade request and waiting for a response). 
The returned `*websocket.Conn` isn't anymore affected by the context.

So the default ensures that the websocket can be established during this time.